### ISSUE TITLE
What pixels are predicted as the extra class Y+1?

### ISSUE THREAD

**Issue by danzhang0615** [tianyu0207/PEBAL#8]
Hi,

I am interested in which kind of pixels are predicted as the extra class after training? In the paper, it appears me that the anomaly prediction is mainly based on the energy score. Thus, I am very curious about the extra class the model has learned to predict. 

Also, how the deep baseline gambler is implemented in the paper to get the numbers in Tab. 1 and 3? Does it use the outlier dataset? what is the reward if not having the energy function? 

Thanks! Looking forward to your reply.

**Comment by yyliu01** [comment]
Thanks for your interest.

For your questions:

1). Which kind of pixels are predicted as the extra class after training? 
We utilise the energy of the inlier classes to judge whether the pixel belongs to normal or not.

2). The extra class the model has learned to predict or not?
The purpose of the extra class in pebal is not used to make prediction, but let the model aware the anomalous pixel during the training. In the original pre-trained model, the inlier pixel will be easy to be recognised (e.g., larger logits in the related channel, and lower energy) and outlier pixel will be hard to be recognised (e.g., lower logits and larger energy). One of the issues is that,  sometimes the outlier is overfitted to the head categories (of the inlier class), while the extra class (in gambler loss) can alleviate it. The gambler loss adds the confident of the extra channel into all the inlier results during the training, which calibrated the inlier entropy/energy indirectly.

For the efficiency of the extra channel, we've tried to utlise entropy-maximisation as the base line (with outlier exposure), but the result is not very impressive. You are very welcome to re-implement it.

3). How the deep baseline gambler is implemented in the paper to get the numbers in Tab. 1 and 3? Does it use the outlier dataset? What is the reward if not having the energy function?
The Deep Gambler is implemented without all other contributions in our approach, and yes, it uses the outlier dataset.  The reward is set to be 4.5 for both inlier and outlier predictions. 

Regards,
Yuyuan

**Comment by danzhang0615** [comment]
Thanks for your quick reply.
1 & 2, yes, I understand for the purpose of outlier segmentation the extra class does not serve as the best proxy, likely worse than the energy score. However, the intention of my question here is more about if you have ever visualized the prediction label map including the extra class. I am quite curious to the observations there, e.g., would the boundary pixels (which are high uncertainty but inlier pixels) be mapped to the extra class?

3) so for the deep gambler, you basically keep the pal loss in your eq. 2 (for both the inliers and outliers), and leaving out the ebm and reg part. Also for the reward, it is a constant setting to 4.5. Is my understanding correct?

**Comment by yyliu01** [comment]
No, we did not visualise the prediction that included the extra channel, but only the energy map during the validation. Please note that we only utilise the energy panel to classify whether it belongs to normal or not, instead of the hard pseudo-labelling (MLE).

For your question:

1 **would the boundary pixels (which are high uncertainty but inlier pixels) be mapped to the extra class?**

The boundary pixels are indeed under high uncertainty, while you can find that most discriminative regions of the anomaly are more distinguishable.

I randomly capture 2 energy outputs of the Fishyscapes below, based on the supported ckpt:

a). Static
![image](https://user-images.githubusercontent.com/102338056/178444690-13a569de-06a4-4d6a-b91d-c578233f38a4.png)

b). L & F
![image](https://user-images.githubusercontent.com/102338056/178444456-9365893a-c7c3-4c4a-946e-3f00c99ef174.png)

With a threshold, it is possible to classify the hard inlier you mentioned and the outlier .

2 Yes, you are correct.

Regards,
Yuyuan

**Comment by danzhang0615** [comment]
Thanks for the visual results. my last question (hopefully :)) to 2. So, for your deep gambler, while emb loss is not a part of it, the score used for anomaly segmentation is still the energy score, which is logsumexp of the 19 classes of CS?

**Comment by yyliu01** [comment]
You are welcome!

The measured way for that should be entropy, following the meta-ood in [here](https://github.com/robin-chan/meta-ood/blob/8657a5e748803998351caa23d44bb95a40b6e32f/evaluation.py#L62). 

If I remember correctly, at that time, the project was not related to energy yet.

**Comment by yyliu01** [comment]
Closing the issue, but feel free to reopen.

**Comment by wmnj** [comment]
> 

Thank you for your explaination,it is very useful for me!!!
And,could you tell me how to get the outlier map you shared with a energy score?
the OE dataset in this paper is mainly used for learning the parameter aw in formula (3)?
![image](https://user-images.githubusercontent.com/33375612/188043445-a191cd54-a3a0-407a-830a-bc4508ac4c7b.png)
